### DISCUSSION TITLE
ImportError: cannot import name 'login_to_hub' from 'doctr.models'

### DISCUSSION THREAD

**Discussion by javohirss** [mindee/doctr#1067] [Q&A]
Hello, I ran into an error when trying to train a model on a custom dataset:
![image](https://user-images.githubusercontent.com/61548557/191177735-86719ce8-a8f9-4248-879e-449b77ae902f.png)

could you help me with this pls

**Comment by felixdittrich92** [comment]
Hi @javohirss :wave:,
seems you have no DL framework installed you can install it the following way:
```
!git clone https://github.com/mindee/doctr.git
%cd doctr
!pip install -e .[tf] --upgrade
!pip install -r references/requirements.txt

import doctr
print(doctr__version__)  <-- should be 0.5.2a0
```

**Reply by felixdittrich92** [reply]
ok there is still a problem with installing doctr on Colab ref: #1036 i think the next release will fix this issue
@frgfm @charlesmindee

**Reply by javohirss** [reply]
Oh, what would you advise me in this case? Sorry, but I'm pretty new to this stuff, but I urgently need to use your tool :)

**Reply by felixdittrich92** [reply]
Mh train the model on your local machine is not possible or no option ?

**Reply by javohirss** [reply]
To be honest, I had no experience with deep learning models on a local machine. As far as I understand, here you need to use the GPU, but if you recommend me to do this, then I will try.

**Reply by felixdittrich92** [reply]
Yep what you need is NVIDIA GPU for local training. Then i would suggest to install conda (https://www.anaconda.com/) . If you not prefer a DL backend i would also suggest to use PyTorch (easy installation with conda: https://pytorch.org/get-started/locally/ )

After this steps install doctr (main branch) on your machine and run your training like in colab

**Reply by javohirss** [reply]
Hello, sorry, I have been installing DLF since yesterday (I chose tensorflow). Then I downloaded your project, went into it through PyCharm. I typed in the terminal those commands that I run in the collab and got the following error: 
![image](https://user-images.githubusercontent.com/61548557/191517318-618ebad8-fddc-4567-9f31-fd339d189453.png)

**Reply by felixdittrich92** [reply]
Hi :wave:, on WIndows you have to follow these steps to install weasyprint: https://doc.courtbouillon.org/weasyprint/stable/first_steps.html#windows

**Reply by javohirss** [reply]
It's me again )
Now with this error:
![image](https://user-images.githubusercontent.com/61548557/191563420-7c34ad72-ee76-4c24-9a0f-c0bebe5629fc.png)

I know it's because of my letters (I'm using Tajik which differs from Russian by 5 characters), but how can I fix it?

**Reply by felixdittrich92** [reply]
Could you try to load the json outside of doctr and check if the same error raised ?

**Reply by felixdittrich92** [reply]
I think you should check that the default encoding in Windows is utf-8 or utf-16

**Comment by javohirss** [comment]
hi, i got same error when I running the json file outside of doctr. There I fixed it by adding utf8 encoding when opening:
![image](https://user-images.githubusercontent.com/61548557/191576939-d4f1dad9-7d51-4f10-98fd-84e6181e5f93.png)
and there is my Windows default encoding:
![image](https://user-images.githubusercontent.com/61548557/191577490-f61579e3-b2d9-4290-915c-517a139d1fcb.png)
Should I add something to your code in this case or change the windows encoding? Btw, thx for paying attention to me)

**Reply by felixdittrich92** [reply]
Yeah as expected i would personally prefer to change the windows encoding :) otherwise you could also change Line 40 in doctr\datasets\recognition.py

**Reply by felixdittrich92** [reply]
No problem happy to help 👍 :)

**Reply by javohirss** [reply]
Hi, it's me again)
Now with this error:
![image](https://user-images.githubusercontent.com/61548557/191741770-f66a3b6b-5d6c-4e12-9d17-ee669bdaf218.png)
I think I installed tensorflow wrong somewhere, that's why it gives this error.
What would you advise me to do?

**Reply by felixdittrich92** [reply]
Could you please try to run:
python references\recognition\train_tensorflow.py crnn_vgg16_bn

Does it raise the same error ?

**Reply by javohirss** [reply]
Sorry, You mean without train_path, val_path, epochs? Cause I already running this (the first line in the picture)

**Reply by felixdittrich92** [reply]
yep only provide the model name

**Reply by javohirss** [reply]
It gaves me this:
![image](https://user-images.githubusercontent.com/61548557/191747115-8077847c-7d34-4466-bff4-e6e2f23927a7.png)

**Reply by felixdittrich92** [reply]
ok then try python references\recognition\train_tensorflow.py crnn_vgg16_bn --font  pass a ttf which is available in your Windows machine  i think Arial.ttf  is in every windows machine available 
so do the following:
`python references\recognition\train_tensorflow.py crnn_vgg16_bn --font Arial.ttf`

**Reply by javohirss** [reply]
Same error, but instead of FreeMono.ttf there is Arial. Should I download some font and load it to some folder?

**Reply by felixdittrich92** [reply]
mh maybe try this: https://stackoverflow.com/questions/44358468/unable-to-use-windows-fonts-with-pil 
or modify the training script: (pass None instead of fonts)
https://github.com/mindee/doctr/blob/e538cc28e9d9e540e418b154420ac0113a1839b2/references/recognition/train_tensorflow.py#L168
https://github.com/mindee/doctr/blob/e538cc28e9d9e540e418b154420ac0113a1839b2/references/recognition/train_tensorflow.py#L256

**Comment by javohirss** [comment]
Hi! This time I ran into this:
![image](https://user-images.githubusercontent.com/61548557/192264784-4e7cfc3e-8bb4-4c8f-baa1-40d81fe383cd.png)

**Reply by felixdittrich92** [reply]
Hi @javohirss :wave:,

`python references\recognition\train_tensorflow.py crnn_vgg16_bn --font Arial.ttf ` was running well ? 
I try to identify if anything with your instalation is wrong or with your data :thinking: 

Would also be good to share some information !
Run : https://github.com/mindee/doctr/blob/main/scripts/collect_env.py  and post the output please :)

**Reply by javohirss** [reply]
I got this:
![image](https://user-images.githubusercontent.com/61548557/192270893-59cbd872-bc95-4356-8035-1fb063eb0b69.png)

**Reply by javohirss** [reply]
and this, after running `python references\recognition\train_tensorflow.py crnn_vgg16_bn --font Arial.ttf` (I have set the None value to font field as you recommended):
![image](https://user-images.githubusercontent.com/61548557/192271500-f1dfbb52-fe52-4294-bce8-3db14b8c34b1.png)

**Reply by felixdittrich92** [reply]
mh really weird but what i see is that there is no NVIDIA driver displayed could you check this please ?

**Comment by javohirss** [comment]
Hi! I don't know why collect_env doesn't show the NVIDIA driver version. But I checked it in GeForce Experience:
![image](https://user-images.githubusercontent.com/61548557/192500351-40fd71ed-58a8-47c7-894f-da118e748700.png)

**Reply by felixdittrich92** [reply]
Hi @javohirss :wave:, i'm not 100% sure but i think with your TF installation is anything not working correctly :thinking: 
Unfortunately without a Windows env i can only suggest steps:

I highly suggest to use Anaconda: https://www.anaconda.com/

TF well working installation: https://conda-forge.org/blog/posts/2021-11-03-tensorflow-gpu/
PyTorch: `conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch`

**Reply by javohirss** [reply]
Hi! Sorry, it's me again. This time I tried to use anaconda as you recommended. I installed tensorflow and tensorflow-gpu (I checked it in Anaconda Navigator) and got the error with which we started our discussion:
![image](https://user-images.githubusercontent.com/61548557/192779313-410d197b-b3bf-4584-be11-4cb881275885.png)
You have recommended me that time to install DL framework with following commands, but how can I do it with conda?

**Reply by felixdittrich92** [reply]
ok with conda:
`conda install tensorflow==2.8.1=cuda112 -c conda-forge`  (this will take some time) - please ensure that all tensorflow installations are deleted before installing

then open a terminal inside doctr and run:
pip install -e .

then run:
conda list tensorflow
conda list python-doctr

please post the output :)

Let me share how it looks on my TF env:
```
(doctr-dev-tf) felix@felix-GS66-Stealth-11UH:~/Desktop/memoresa-work-ml$ conda list tensorflow
# packages in environment at /home/felix/.conda/envs/doctr-dev-tf:
#
# Name                    Version                   Build  Channel
tensorflow                2.8.1           cuda112py38hded6998_0    conda-forge
tensorflow-addons         0.16.1                   pypi_0    pypi
tensorflow-base           2.8.1           cuda112py38had2df90_0    conda-forge
tensorflow-estimator      2.8.1           cuda112py38hf5dcc89_0    conda-forge
(doctr-dev-tf) felix@felix-GS66-Stealth-11UH:~/Desktop/memoresa-work-ml$ conda list python-doctr
# packages in environment at /home/felix/.conda/envs/doctr-dev-tf:
#
# Name                    Version                   Build  Channel
python-doctr              0.5.2a0                   dev_0    <develop>
```

**Reply by javohirss** [reply]
Hi!
![image](https://user-images.githubusercontent.com/61548557/192846444-a647b337-8ac4-45dd-8c24-ccb65f1dcf54.png)
Sorry, I ran into a problem installing with conda-forge version 2.8.1, and I reinstalled tensorflow as written on the official site.
Is everything fine here?

**Reply by felixdittrich92** [reply]
Mh ... unfortunately i cannot validate this remote :/ (if it installed with conda i know that it is correctly) what was the problem with conda ?

**Reply by felixdittrich92** [reply]
But doctr is installed correct 👍

**Reply by javohirss** [reply]
![image](https://user-images.githubusercontent.com/61548557/192855489-a0ec4ab2-9d04-4e69-a16c-a07e60ac791d.png)
This what I got when I was trying to install with conda-forge

**Reply by javohirss** [reply]
I think this installation doesn't work on Windows:
from [https://conda-forge.org/blog/posts/2021-11-03-tensorflow-gpu/](url)
![image](https://user-images.githubusercontent.com/61548557/192945887-024e4a37-fdaa-4a82-9932-c7f508ab5940.png)

**Reply by felixdittrich92** [reply]
Hi @javohirss :wave:, yes sry you are right after taking a look at conda-forge each arch has a build without windows :sweat_smile:.
Have you tested if it works now as is ? 
So at the end `python references\recognition\train_tensorflow.py crnn_vgg16_bn --font Arial.ttf` has to work if the font is installed and can be loaded with Pillow ... :/

**Reply by javohirss** [reply]
Hi! Sorry, as You predicted, I have a font problem :sweat_smile: . 
![image](https://user-images.githubusercontent.com/61548557/193007418-cd74c903-ca19-44c6-8273-dc4f414d5d6c.png)

You said to load it so that Pillow can load it. Could you tell me how to do it, I mean upload it somewhere in windows or do it through python?

**Reply by felixdittrich92** [reply]
Hi :wave:,

Download font: (don't forget to unzip the folder)
https://fontmeme.com/schriftarten/freemono-schriftart/

Install font:
https://www.groovypost.com/howto/install-fonts-windows-10/

test:
```
from PIL import ImageFont

font = ImageFont.truetype("FreeMono.ttf", 32)
```

then test:
`python references\recognition\train_tensorflow.py crnn_vgg16_bn --font FreeMono.ttf`

If the font can be found and your TF installation is correct this will run :hugs:

**Reply by javohirss** [reply]
Hi, sorri, I installed font and checked it with PIL code as you said, but I still get same error:
![image](https://user-images.githubusercontent.com/61548557/193264877-a8eef6be-0726-4ea5-ae46-90617ceeec0b.png)

**Reply by felixdittrich92** [reply]
Hi @javohirss :wave:,
loading FreeMono.ttf works well ? 
And this error occurs by running: `python references\recognition\train_tensorflow.py crnn_vgg16_bn --font FreeMono.ttf` ?
If so let's try to force it with modyfing:
https://github.com/mindee/doctr/blob/3d7c8127bc8df8e28ab28981474d291446df06d4/doctr/datasets/generator/base.py#L41
to: 
`text_w, text_h = font.getsize(text.encode('utf-8'))`

**Reply by javohirss** [reply]
Now:
![image](https://user-images.githubusercontent.com/61548557/193270267-dbdd968e-7035-4798-a5ab-4ba3f2c67040.png)

**Reply by felixdittrich92** [reply]
I don't understand why it shows the font warning ... if you say it loads without errors :thinking:

**Reply by felixdittrich92** [reply]
@frgfm any idea how we can help remotely better with this ? :sweat_smile:

**Reply by javohirss** [reply]
If it helps in any way: I tried running with PyTorch and got this:
![image](https://user-images.githubusercontent.com/61548557/193796685-6026226f-e7c5-4325-abd0-f28abf3da3ae.png)
![image](https://user-images.githubusercontent.com/61548557/193796962-39118d85-59ce-4d47-bd4a-7c069478bb14.png)
Now I will try to run it on Linux

**Reply by felixdittrich92** [reply]
Hi @javohirss :wave:,
have you setup PyTorch in a new env how i wrote in the email (with python 3.8) ?
If i remember correctly we have had a while ago an similar problem with python 3.10 on windows machines, where downgrading to python 3.8 fixed the problem :)

**Reply by javohirss** [reply]
I tried with python 3.8.12 and still got:
![image](https://user-images.githubusercontent.com/61548557/193814981-2a051ffc-e89a-46d2-8dc7-b6e14297c556.png)

**Reply by felixdittrich92** [reply]
Mh yeah still some issues with multiprocessing on Windows :/ 
Replacing fastprogress with tqdm and setting num_workers=0 (dataloader) should help.. but if possible switch to Linux is the better alternative 😅

**Reply by javohirss** [reply]
Hello! I switched to Linux and ran the code:
`python references\recognition\train_pytorch.py crnn_vgg16_bn --font FreeMono.ttf`
And it finally worked!
Then I added tajik lang (it differs from Russian in the presence of several characters) to vocabs.py:
![ksnip_20221009-002835](https://user-images.githubusercontent.com/61548557/194724267-20560849-7dd1-4a34-9d96-65b7b6474ee1.png)
and got error:

![ksnip_20221009-003029](https://user-images.githubusercontent.com/61548557/194724311-248c2d2c-9131-419d-8b8f-427164ea0943.png)

**Reply by felixdittrich92** [reply]
Hi @javohirss 👋 ,
good news 👍 
This says that there are characters in your dataset which are not present in the given vocab please verify this :) 

For example one label: 'xyz' and you provide as vocab: 'xy'

**Reply by javohirss** [reply]
Hello! At first, it really turned out that there were characters in the dataset that were not contained in the vocab. I wrote a function that checks each label against a vocab to see if the label's characters are in the vocab. I have corrected all wrong characters but still getting this error. Is there any way to find exactly where something is going wrong? I mean that wrong label or character.

**Comment by felixdittrich92** [comment]
Hi copy the function: https://github.com/mindee/doctr/blob/3d7c8127bc8df8e28ab28981474d291446df06d4/doctr/datasets/utils.py#L62
and test against your labels.json with a short script so you can easily find the samples which raise the error :)

**Reply by javohirss** [reply]
Hi! It finally worked. I just set --vocab "tajik", cause by deafult it' was "french".
I have trained recognition model and now I have got a archive with pickle file and data.
What shall I do now?

**Reply by felixdittrich92** [reply]
Now you can load your trained model:

See answer #940 
For TF use `model.load_weights('path to folder/weights')`

**Reply by javohirss** [reply]
`model = crnn_vgg16_bn(pretrained=False, vocab=VOCABS["french"])`
`model.load_state_dict('path, mapping='cpu')`
`predictor = ocr_predictor(reco_arch=model, pretrained=True)`
Sorry, I don't get, where shall I run it..? 
and path..path to my trained model archive or to pckl file..?

**Reply by felixdittrich92** [reply]
In your inference script like:

```
doc = DocumentFile.from_images(..)
model = crnn_vgg16_bn(pretrained=False, vocab=VOCABS['tajik'])
model.load_weights('path/weights')
predictor = ocr_predictor(reco_arch=model, pretrained=True)
result = predictor(doc)
print(result)
```

After training with TF there is a folder generated which contains the weights, etc. give the full path + /weights

**Reply by javohirss** [reply]
I trained it with PyTorch, and there is only a folder crnn_vgg16_bn_20221011.pt, that contains data, pickle file and version:

![ksnip_20221011-180121](https://user-images.githubusercontent.com/61548557/195098004-431f42fb-b376-4b1a-9fb6-08a1f7748833.png)

Maybe I am wrong, but I really cannot find weights path, is a pickle file trained model?

**Reply by felixdittrich92** [reply]
Then you can use the example from #940 and the path to .pt file -> xyz/model.pt

**Reply by felixdittrich92** [reply]
Hi @javohirss does it work now for you ? :)

**Reply by javohirss** [reply]
Hi !:grin: 
Yep, finally I ran it, but I got a very bad results, the model could not even recognize the numbers.
Shall I train it for more epochs (I have trained it for 5 epochs) or something else..?
Ah, and one more thing, I need to recognize english at the same time too. How should I do it? :smile:

**Reply by felixdittrich92** [reply]
Hi @javohirss :wave:,
you will need lots of data to do this (english and tajik samples)
and you need to modify the training script:  

NOTE: while inference you have to provide the vocab in the same way

```
from collections import OrderedDict
vocab = ''.join(OrderedDict.fromkeys(VOCABS['tajik'] + VOCABS['english'])    
```

About the training i need more details about the logs (loss / exact / partial) and maybe 1 or 2 samples from your dataset

**Reply by javohirss** [reply]
Logs:
![Screenshot from 2022-10-12 16-29-00](https://user-images.githubusercontent.com/61548557/195331458-4505d741-d3e4-4985-a0c5-47573dc15914.png)

First I tested on passport image (that's what I need in my project) and got bad results, but I thought that it was in the presence of English, and I used an image consisting entirely of Tajik. Result is also bad :

![Screenshot from 2022-10-12 16-27-34](https://user-images.githubusercontent.com/61548557/195332059-94f05ee7-1a7a-465a-acff-dc0ce0dbacd0.png)

My dataset consists of 20k train images and 5k val. Examples:

![АБДУРАЗЗОҚ_24261](https://user-images.githubusercontent.com/61548557/195332447-e10dfa92-fcc4-4d32-b737-38e9eadd3c3f.jpg)

![Эвин_10344](https://user-images.githubusercontent.com/61548557/195332533-4433f937-930f-4d20-bb99-b5aa52be4178.jpg)



Maybe it's all about fonts?

**Reply by felixdittrich92** [reply]
Mh yes, to compare the pretrained versions was trained on ~10M real world word crops. 
If you train the model from scratch you need much more data (with different fonts, backgrounds, etc.). 
I would also suggest (if possible) to use as real world dataset to validate the model (so in short train it with synthetic data and use a much smaller dataset in --val_path then the metrics are much more meaningful)

**Reply by javohirss** [reply]
Ahh.. it's Tajik. I am pretty sure that there is no such dataset 😅 
Theoretically, I could create it myself 🙊 , but how many images do we need? I mean for validation
And is the given script in the _generator_ folder for data synthesis?

**Reply by felixdittrich92** [reply]
Hi @javohirss :wave:,
Yeah unfortunately i think too :/
You have to test a bit with it but i would say for real word crops a range between 5k - 10k should be possible and a good starting point.
Our word generator needs some improvements but sure you could test it.
Example:
`python3 /home/user/Desktop/doctr/references/recognition/train_pytorch.py crnn_vgg16_bn --train-samples 10000 --val-samples 1000 --min-chars 1 --max-chars 25  --font xyz.ttf `

change:
https://github.com/mindee/doctr/blob/acb9f64b11ebad8e53ac60737fcde8dbd3158a22/references/recognition/train_pytorch.py#L185
to:
```
from collections import OrderedDict
vocab = ''.join(OrderedDict.fromkeys(VOCABS['tajik'] + VOCABS['english'])  
```  

This creates automatically a synthetic dataset with the generator and triggers the training 
ATTENTION: you need to install fonts which are compatible with tajik and english in your case
more information about installing custom fonts on your machine:
https://github.com/mindee/doctr/tree/main/references/recognition#using-custom-fonts

Offtopic: https://github.com/clovaai/synthtiger could also be a very good option to create a dataset 

If you have the possibility to use your GPU for the inference and latency is of secondary importance, you could also try our MASTER implementation, i.e. look at the model. This generalizes much faster to less data. I was able to achieve good results with it even with toy experiments.

Cheers :v:
